### ISSUE TITLE
V2 Pre-Token Generation Lamba / Cognito Support

### ISSUE THREAD

**Issue by chris-patten-parsable** [serverless/serverless#12336]
### Is there an existing issue for this?

- [X] I have searched existing issues, it hasn't been reported yet

### Use case description

AWS has release a V2 feature and would like SLS to support it.  (https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-pre-token-generation.html).  It appears that the current solution is hardcoded to V1 (https://github.com/serverless/serverless/blob/7b1e0120b0c97e811ea32b67a51e4c9e9ccf4edd/lib/plugins/aws/package/compile/events/cognito-user-pool.js#L23).  Not sure what it takes to get this feature working but it would be nice to have the support!  Looks like CloudFrount exposes in the Cognito Config -> Lambda Config -> PreTokenGenerationConfig -> LambdaVersion.  Hope that helps!

**Comment by btamadio** [comment]
I'm also blocked by this right now.

Unfortunately, it seems the line you've posted here doesn't control the allowed versions for PreTokenGeneration events, as it only gets evaluated when the trigger is one of "CustomEmailSender" or "CustomSMSSender"

See:
https://github.com/serverless/serverless/blob/7b1e0120b0c97e811ea32b67a51e4c9e9ccf4edd/lib/plugins/aws/package/compile/events/cognito-user-pool.js#L189

**Comment by jrdnull** [comment]
I was able to do this by specifying the LambdaConfig in the UserPool resource, e.g

```yml
    DodsUserPool:
      Type: AWS::Cognito::UserPool
      Properties:
        UserPoolAddOns:
          AdvancedSecurityMode: "AUDIT"
        LambdaConfig:
          PreTokenGenerationConfig:
            LambdaArn: arn:aws:lambda:${aws:region}:${aws:accountId}:function:{$self:service}-${self:provider.stage}-preTokenGeneration
            LambdaVersion: V2_0
        ...
```

Also specify permission else it won't be able to invoke it

```yml
    PreTokenGenerationLambdaInvokePermission:
      Type: "AWS::Lambda::Permission"
      Properties:
        FunctionName: !GetAtt PreTokenGenerationLambdaFunction.Arn
        Action: "lambda:InvokeFunction"
        Principal: "cognito-idp.amazonaws.com"
```

**Comment by czubocha** [comment]
Shipped in **[v4.37.0](https://github.com/serverless/serverless/releases/tag/sf-core%404.37.0)**.

The `cognitoUserPool` `PreTokenGeneration` trigger now accepts a `lambdaVersion` property, so it's no longer hardcoded to V1:

```yaml
functions:
  preTokenGeneration:
    handler: preToken.handler
    events:
      - cognitoUserPool:
          pool: MyUserPool
          trigger: PreTokenGeneration
          lambdaVersion: V2_0
```

Accepted values are `V1_0`, `V2_0` (ID + access token customization), and `V3_0` (V2 plus machine-to-machine client-credentials grants). When omitted, the emitted CloudFormation is byte-identical to before, so existing services see no diff on upgrade. Note that V2_0/V3_0 require the Cognito Essentials or Plus feature plan.

Docs: https://www.serverless.com/framework/docs/providers/aws/events/cognito-user-pool#pretokengeneration-trigger

Closing as released - thanks for the request!
